js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,575 @@
+	errc chan<- error
+}
+
+func (t transportCreds) ClientHandshake(ctx context.Context,
+	addr string, in net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	out, auth, err := t.TransportCredentials.ClientHandshake(ctx, addr, in)
+	if err != nil {
+		t.errc <- err
+	}
+
+	return out, auth, err
+}
+
+// Connect is a block dial to the gRPC server at the given address (host:port)
+// nolint: funlen
+func (c *Client) Connect(ctxPtr *context.Context, addr string, params map[string]interface{}) (bool, error) {
+	state := lib.GetState(*ctxPtr)
+	if state == nil {
+		return false, errConnectInInitContext
+	}
+
+	isPlaintext, timeout := false, 60*time.Second
+
+	for k, v := range params {
+		switch k {
+		case "plaintext":
+			isPlaintext, _ = v.(bool)
+		case "timeout":
+			var err error
+			switch t := v.(type) {
+			case string:
+				if timeout, err = time.ParseDuration(t); err != nil {
+					return false, fmt.Errorf("unable to parse %q: %v", k, err)
+				}
+			case int64:
+				timeout = time.Duration(float64(t)) * time.Millisecond
+			case float64:
+				timeout = time.Duration(t) * time.Millisecond
+			default:
+				return false, fmt.Errorf("unable to use type %T as a timeout value", v)
+			}
+		default:
+			return false, fmt.Errorf("unknown connect param: %q", k)
+		}
+	}
+
+	// (rogchap) Even with FailOnNonTempDialError, if there is a TLS error this will timeout
+	// rather than report the error, so we can't rely on WithBlock. By running in a goroutine
+	// we can then wait on the error channel instead, which could happen before the Dial
+	// returns. We only need to close the channel to un-block in a non-error scenario;
+	// otherwise it can be GCd without closing as we return on an error on the channel.
+	errc := make(chan error, 1)
+	go func() {
+		opts := []grpc.DialOption{
+			grpc.WithBlock(),
+			grpc.FailOnNonTempDialError(true),
+			grpc.WithStatsHandler(c),
+		}
+
+		if ua := state.Options.UserAgent; ua.Valid {
+			opts = append(opts, grpc.WithUserAgent(ua.ValueOrZero()))
+		}
+
+		if !isPlaintext {
+			tlsCfg := state.TLSConfig.Clone()
+			tlsCfg.NextProtos = []string{"h2"}
+
+			// TODO(rogchap): Would be good to add support for custom RootCAs (self signed)
+
+			// (rogchap) we create a wrapper for transport credentials so that we can report
+			// on any TLS errors.
+			creds := transportCreds{
+				credentials.NewTLS(tlsCfg),
+				errc,
+			}
+			opts = append(opts, grpc.WithTransportCredentials(creds))
+		}
+
+		if isPlaintext {
+			opts = append(opts, grpc.WithInsecure())
+		}
+
+		dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+			return state.Dialer.DialContext(ctx, "tcp", addr)
+		}
+		opts = append(opts, grpc.WithContextDialer(dialer))
+
+		ctx, cancel := context.WithTimeout(*ctxPtr, timeout)
+		defer cancel()
+
+		var err error
+		c.conn, err = grpc.DialContext(ctx, addr, opts...)
+		if err != nil {
+			errc <- err
+
+			return
+		}
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
```

Comment:
Would canceling the ctx ... make the goroutine return faster?
Maybe we can drop the goroutine ... have the `transportCreds` remember the first error they get and cancel the context on it?

Reply:
Possibly. We wound have to attach a context cancel function to the `transportCreds` object at the beginning, and in the event of an error call `cancel()`, and then store the TLS error on the `transportCreds` to later retrieve the error to check if the error was caused by TLS (as appose to some other context cancellation).

In my opinion this last part of checking an error field on a struct is not idiomatic Go and is prone to being missed (hence why I don't like the builder pattern in Go, you have to remember to check that error field). 
At this point I'm still in-favour of the error channel an goroutine as I think it reads cleaner.